src/lib/modelUtils.ts

```diff
@@ -0,0 +1,284 @@
+  data: Record<string, any>,
+  model,
+): Promise<{
+  matched: Record<string, any>,
+  unmatched: Record<string, any>,
+}> => {
+  // Retrieve the column information for the model
+  const modelData = await getColumnInformation(model);
+
+  return Object.entries(data)
+    .reduce((acc, [key, value]) => {
+      // Find the matching column in the model data
+      const matchColumn = modelData.find((md) => md.columnName === key);
+      if (matchColumn
+        && ((value === null && matchColumn?.allowNull)
+          || (typeof value === dataTypeMapping[matchColumn?.dataType?.key]))
+      ) {
+        // If the value matches the column criteria, add it to the matched object
+        acc.matched[key] = value;
+      } else {
+        // Otherwise, add it to the unmatched object
+        acc.unmatched[key] = value;
+      }
+
+      return acc;
+    }, { matched: {}, unmatched: {} });
+};
+
+/**
+ * This function includes a model and its associated data based on the provided include function.
+ * It retrieves all records that match the given conditions.
+ *
+ * @param includeFunc - The function that defines the model and its associations to be included.
+ * @param moreWhere - Additional conditions to filter the records.
+ * @param funcArgs - Arguments to be passed to the include function.
+ * @param attributes - Attributes to be selected from the records.
+ * @returns A promise that resolves to an array of matching records.
+ */
+const includeToFindAll = async (
+  includeFunc,
+  moreWhere,
+  funcArgs = null,
+  attributes = null,
+) => {
+  // Destructure the properties 'as', 'model', 'where', and other arguments from the result of
+  // calling the include function
+  const {
+    as,
+    model,
+    where,
+    ...args
+  } = funcArgs
+    ? includeFunc(...funcArgs)
+    : includeFunc();
+
+  // Find all records of the model, applying the merged 'where' condition and any additional
+  // conditions
+  return model.findAll({
+    where: merge(where, moreWhere),
+    ...(attributes && { attributes }),
+    ...args,
+  });
+};
+
+/**
+ * Transforms a nested object or array of objects by recursively stripping out
+ * Sequelize model instance metadata and retaining only the raw data values.
+ * This function is intended to be used with raw Sequelize query results.
+ *
+ * @param data - An object or array of objects potentially containing Sequelize
+ *               model instance properties such as `dataValues` and metadata.
+ *
+ * @returns A new object or array of objects similar to the input `data` but
+ *          with Sequelize-specific properties removed. The structure of the
+ *          input is preserved, and only the raw data values are retained.
+ *
+ * @remarks
+ * This function assumes that Sequelize model instances have a `dataValues`
+ * property containing the raw data, as well as other Sequelize-specific
+ * metadata properties. These metadata properties are excluded from the output.
+ */
+const nestedRawish = (
```

Comment:
Hate to bring this up, but is this different than .toJSON() ?

Reply:
findOne: nestedRawish(result) === result.toJSON()
findAll: nestedRawish(result) === result.map((r) => r.toJSON())

So I guess its very vary similar, it just handles and array at the top level that .toJSON does not.

Reply:
I don't have a way to quickly test virtual columns, So I don't know if .toJSON() handles that case. nestedRawish does not right now.

Reply:
toJSON is the way to get virtual columns in a plain output, where raw: true does not

Not meant to be a criticism of nestedRawish either, I was more trying to understand when to use which